tests/OmniSharp.Roslyn.CSharp.Tests/CustomRoslynAnalyzerFacts.cs

```diff
@@ -0,0 +1,233 @@
+            }
+        }
+
+        [DiagnosticAnalyzer(LanguageNames.CSharp)]
+        public class TestDiagnosticAnalyzer : DiagnosticAnalyzer
+        {
+            public TestDiagnosticAnalyzer(string id, bool isEnabledByDefault)
+            {
+                this.id = id;
+                _isEnabledByDefault = isEnabledByDefault;
+            }
+
+            private DiagnosticDescriptor Rule => new DiagnosticDescriptor(
+                this.id,
+                "Testtitle",
+                "Type name '{0}' contains lowercase letters",
+                "Naming",
+                DiagnosticSeverity.Warning,
+                isEnabledByDefault: _isEnabledByDefault
+            );
+
+            private readonly string id;
+            private readonly bool _isEnabledByDefault;
+
+            public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            {
+                get { return ImmutableArray.Create(Rule); }
+            }
+
+            public override void Initialize(AnalysisContext context)
+            {
+                context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+            }
+
+            private void AnalyzeSymbol(SymbolAnalysisContext context)
+            {
+                var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+                if (namedTypeSymbol.Name == "_this_is_invalid_test_class_name")
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        Rule,
+                        namedTypeSymbol.Locations[0],
+                        namedTypeSymbol.Name
+                    ));
+                }
+            }
+        }
+
+        public CustomRoslynAnalyzerFacts(ITestOutputHelper output, SharedOmniSharpHostFixture sharedOmniSharpHostFixture)
+            : base(output, sharedOmniSharpHostFixture)
+        {
+        }
+
+        [Fact]
+        public async Task When_custom_analyzers_are_executed_then_return_results()
+        {
+            var testFile = new TestFile("testFile.cs", "class _this_is_invalid_test_class_name { int n = true; }");
+            var codeCheckService = GetRequestHandler(SharedOmniSharpTestHost);
+
+            SharedOmniSharpTestHost.AddFilesToWorkspace(testFile);
+
+            var testAnalyzerRef = new TestAnalyzerReference("TS1234");
+
+            var projectIds = CreateProjectWitFile(testFile, testAnalyzerRef);
+
+            // This retry should be replaced with emitted events listener aproach after LSP is mainstream.
+            // Retry is required when build system greatly slows down some times, another issue is that
+            // it feels that some of update events from workspace get lost and it requires further investigation.
+            // If missing events are true then same issue will happen with LSP version too.
+            await RetryAssert.On<ContainsException>(async () =>
```

Comment:
What would be a better way to fix this? Is there a way to ask the diagnostics service when its queue is empty?

Put differently: when this starts failing in CI, what are we supposed to do as maintainers of the product? Increase the retry count and hope it's not concealing a bug?

Reply:
Removed these retries. It seems problem was omnisharp shared test host side effect between tests.